niworkflows/interfaces/bids.py

```diff
@@ -288,6 +289,493 @@ def _run_interface(self, runtime):
+                    try:
+                        data_dtype = nb.load(self.inputs.source_file[0]).get_data_dtype()
+                    except Exception:
+                        LOGGER.warning(
+                            f"Could not get data type of file {self.inputs.source_file[0]}"
+                        )
+                        data_dtype = None
+
+                if data_dtype:
+                    data_dtype = np.dtype(data_dtype)
+                    orig_dtype = nii.get_data_dtype()
+                    if orig_dtype != data_dtype:
+                        LOGGER.warning(
+                            f"Changing {out_file} dtype from {orig_dtype} to {data_dtype}"
+                        )
+                        # coerce dataobj to new data dtype
+                        if np.issubdtype(data_dtype, np.integer):
+                            new_data = np.rint(nii.dataobj).astype(data_dtype)
+                        else:
+                            new_data = np.asanyarray(nii.dataobj, dtype=data_dtype)
+                        # and set header to match
+                        if new_header is None:
+                            new_header = nii.header.copy()
+                        new_header.set_data_dtype(data_dtype)
+                del nii
+
+            if new_data is new_header is None:
+                out_file = orig_file
+            else:
+                out_file = Path(runtime.cwd) / dest_file.name
+
+                orig_img = nb.load(orig_file)
+                if new_data is None:
+                    set_consumables(new_header, orig_img.dataobj)
+                    new_data = orig_img.dataobj.get_unscaled()
+                else:
+                    # Without this, we would be writing nans
+                    # This is our punishment for hacking around nibabel defaults
+                    new_header.set_slope_inter(slope=1., inter=0.)
+                unsafe_write_nifti_header_and_data(
+                    fname=out_file,
+                    header=new_header,
+                    data=new_data
+                )
+                del orig_img
+
+            self._results["out_file"].append(str(out_file))
+
+        return runtime
+
+
+class _SaveDerivativeInputSpec(TraitedSpec):
+    base_directory = Directory(
+        exists=True, mandatory=True, desc="Path to the base directory for storing data."
+    )
+    in_file = InputMultiObject(
+        File(exists=True), mandatory=True, desc="the object to be saved"
+    )
+    metadata = InputMultiObject(
+        traits.DictStrAny, desc="metadata to be saved alongside the file"
+    )
+    relative_path = InputMultiObject(
+        traits.Str, desc="path to the file relative to the base directory"
+    )
+
+class _SaveDerivativeOutputSpec(TraitedSpec):
+    out_file = OutputMultiObject(File(exists=True), desc="written file path")
+    out_meta = OutputMultiObject(File(exists=True), desc="written JSON sidecar path")
```

Comment:
i remember removing `exists` from outputs because of potential caching issues - has that been confirmed / resolved?

Reply:
Yeah, we should remove the exists check.